parser_combinator.py

```diff
@@ -0,0 +1,286 @@
+import typing as tp
+from typing_extensions import NamedTuple
+import re
+
+
+class ParseError(NamedTuple):
+    msg: str
+
+
+T = tp.TypeVar('T')
+K = tp.TypeVar('K')
+V = tp.TypeVar('V')
+S = tp.TypeVar('S')
+
+ParseResult: tp.TypeAlias = tp.Tuple[T, str] | ParseError
+Parser = tp.Callable[[str], ParseResult[T]]
+
+
+def pmap(p: Parser[T], f: tp.Callable[[T], K]) -> Parser[K]:
+    def fn(s: str) -> ParseResult[K]:
+        maybeVal = p(s)
+        if isinstance(maybeVal, ParseError):
+            return maybeVal
+        (val, s) = maybeVal
+        newVal = f(val)
+        return (newVal, s)
+    return fn
+
+
+def parse(p: Parser[T], s: str) -> ParseResult[T]:
+    maybeVal = p(s)
+    if isinstance(maybeVal, ParseError):
+        return maybeVal
+    (t, s) = maybeVal
+    return (t, s)
+
+
+def char(c: str) -> Parser[str]:
+    def fn(s: str) -> ParseResult[str]:
+        if len(c) != 1:
+            return ParseError("char called with string")
```

Comment:
Shouldn't happen here, this should be an assertion after the function char is declared.